hydra-cluster/test/Test/Hydra/Cluster/HydraClientSpec.hs

```diff
@@ -0,0 +1,294 @@
+    bobKeys@(bobCardanoVk, _) <- generate genKeyPair
+    carolKeys@(carolCardanoVk, _) <- generate genKeyPair
+
+    let cardanoKeys = [aliceKeys, bobKeys, carolKeys]
+        hydraKeys = [aliceSk, bobSk, carolSk]
+
+    let firstNodeId = 1
+    hydraScriptsTxId <- publishHydraScriptsAs node Faucet
+    let contestationPeriod = UnsafeContestationPeriod 2
+    let hydraTracer = contramap FromHydraNode tracer
+    withHydraCluster hydraTracer tmpDir nodeSocket firstNodeId cardanoKeys hydraKeys hydraScriptsTxId contestationPeriod $ \nodes -> do
+      let [n1, n2, n3] = toList nodes
+      waitForNodesConnected hydraTracer 20 $ n1 :| [n2, n3]
+
+      -- Funds to be used as fuel by Hydra protocol transactions
+      seedFromFaucet_ node aliceCardanoVk 100_000_000 (contramap FromFaucet tracer)
+      seedFromFaucet_ node bobCardanoVk 100_000_000 (contramap FromFaucet tracer)
+      seedFromFaucet_ node carolCardanoVk 100_000_000 (contramap FromFaucet tracer)
+
+      action node nodes hydraTracer
+
+prepareScenario ::
+  RunningNode ->
+  NonEmpty HydraClient ->
+  Tracer IO EndToEndLog ->
+  IO (TxId, HeadId, (VerificationKey PaymentKey, SigningKey PaymentKey), (VerificationKey PaymentKey, SigningKey PaymentKey))
+prepareScenario node nodes tracer = do
+  let [n1, n2, n3] = toList nodes
+  let hydraTracer = contramap FromHydraNode tracer
+
+  send n1 $ input "Init" []
+  headId <-
+    waitForAllMatch 10 [n1, n2, n3] $
+      headIsInitializingWith (Set.fromList [alice, bob, carol])
+
+  -- Get some UTXOs to commit to a head
+  aliceKeys@(aliceExternalVk, aliceExternalSk) <- generate genKeyPair
+  committedUTxOByAlice <- seedFromFaucet node aliceExternalVk aliceCommittedToHead (contramap FromFaucet tracer)
+  requestCommitTx n1 committedUTxOByAlice <&> signTx aliceExternalSk >>= submitTx node
+
+  bobKeys@(bobExternalVk, bobExternalSk) <- generate genKeyPair
+  committedUTxOByBob <- seedFromFaucet node bobExternalVk bobCommittedToHead (contramap FromFaucet tracer)
+  requestCommitTx n2 committedUTxOByBob <&> signTx bobExternalSk >>= submitTx node
+
+  requestCommitTx n3 mempty >>= submitTx node
+
+  let u0 = committedUTxOByAlice <> committedUTxOByBob
+
+  waitFor hydraTracer 10 [n1, n2, n3] $ output "HeadIsOpen" ["utxo" .= u0, "headId" .= headId]
+
+  -- Create an arbitrary transaction using some input to have history.
+  tx <- sendTx nodes committedUTxOByAlice aliceExternalSk bobExternalVk paymentFromAliceToBob
+  waitFor hydraTracer 10 (toList nodes) $
+    output "TxValid" ["transactionId" .= txId tx, "headId" .= headId, "transaction" .= tx]
+  pure (txId tx, headId, aliceKeys, bobKeys)
+
+-- NOTE(AB): this is partial and will fail if we are not able to generate a payment
+sendTx :: NonEmpty HydraClient -> UTxO' (TxOut CtxUTxO) -> SigningKey PaymentKey -> VerificationKey PaymentKey -> Lovelace -> IO Tx
+sendTx nodes senderUTxO sender receiver amount = do
+  let utxo = Prelude.head $ UTxO.pairs senderUTxO
```

Comment:
Is using `head` safe here? Perhaps we can loose some information if UTxO contains more than one entry?

Reply:
head is safe to be used here.

also, this is how we select an output in other e2e scenarios.